coreneuron/sim/fadvance_core.cpp

```diff
@@ -286,6 +286,11 @@ void nrncore2nrn_send_values(NrnThread* nth) {
 
     TrajectoryRequests* tr = nth->trajec_requests;
     if (tr) {
+        // \todo Check if user has requested voltages for this NrnThread object.
+        //       Currently we are updating voltages if there is any trajectory
+        //       requested by NEURON.
+        update_voltage_from_gpu(nth); 
```

Comment:
Although it is true that most requests are for voltage. The present non-gpu implementation allows any RANGE variable and t.  Is the entire data array too much (and isn't that already done at the end of a run anyway?)

Reply:
yes, updating data from gpu to cpu could be too heavy. For efficiency reason we might need to accumulate buffers directly on GPU (based on memory requirements). This advanced implementation need to be done separately.

With this PR, I am finishing the functionality required for bulb-3d model.